Source/JavaScriptCore/runtime/WaiterListManager.cpp

```diff
@@ -0,0 +1,289 @@
+            value = vm.smallStrings.notEqualString();
+        else if (!timeout)
+            value = vm.smallStrings.timedOutString();
+        else {
+            isAsync = true;
+            JSPromise* promise = JSPromise::create(vm, globalObject->promiseStructure());
+            addAsyncWaiter(ptr, promise, timeout);
+            value = promise;
+        }
+
+        object->putDirect(vm, vm.propertyNames->async, jsBoolean(isAsync));
+        object->putDirect(vm, vm.propertyNames->value, value);
+        return object;
+    }
+
+    if (validation == AtomicsWaitValidation::Fail)
+        return vm.smallStrings.notEqualString();
+
+    // Notice that we allocate SyncWaiter on stack, in other words it should be
+    // the only owner for the allocated Waiter ideally (This is not true in the 
+    // WaiterList::take function since there is a protectedWaiter. And at that 
+    // point the refCount will be 2 instead of 1).
+    Waiter syncWaiter { &vm };
+    Ref<WaiterList> list = findOrCreateList(ptr);
+    MonotonicTime time = MonotonicTime::now() + timeout;
+
+    {
+        Locker listLocker { list->lock };
+        list->addLast(listLocker, &syncWaiter);
+        dataLogLnIf(WaiterListsManagerInternal::verbose, "WaiterListManager added a new SyncWaiter ", RawPointer(&syncWaiter), " to a waiterList for ptr ", RawPointer(ptr));
+
+        while (syncWaiter.vm() && time.now() < time)
+            syncWaiter.getCondition().waitUntil(list->lock, time.approximateWallTime());
+
+        // At this point, syncWaiter should be either notified (dequeued) or timeout (not dequeued).
+        // If it's notified by other thread, it's vm should be nulled out.
+        bool didGetDequeued = !syncWaiter.vm();
+        ASSERT(didGetDequeued || syncWaiter.vm() == &vm);
+        if (didGetDequeued)
+            return vm.smallStrings.okString();
+
+        list->take(listLocker, &syncWaiter);
+        return vm.smallStrings.timedOutString();
+    }
+}
+
+void WaiterListManager::addAsyncWaiter(void* ptr, JSPromise* promise, Seconds timeout)
+{
+    Ref<WaiterList> list = findOrCreateList(ptr);
+    Ref<Waiter> waiter = adoptRef(*new Waiter(promise));
+
+    {
+        Locker listLocker { list->lock };
+        list->addLast(listLocker, waiter.ptr());
+
+        if (timeout != Seconds::infinity()) {
+            Ref<RunLoop::DispatchTimer> timer = RunLoop::current().dispatchAfter(timeout, [this, ptr, waiter = waiter.copyRef()] {
+                RefPtr<WaiterList> list = findList(ptr);
+                if (list) {
+                    Locker listLocker { list->lock };
+                    // If the waiter is notified or removed from list by other thread which failed to
+                    // cancel the corresponding timer, we should not notify it.
+                    if (waiter->getTicket())
+                        notifyWaiterImpl(listLocker, list->take(listLocker, waiter.ptr()), ResolveResult::Timeout);
+                }
```

Comment:
This is not correct since waiterlist is removed because SharedArrayBuffer is destroyed, and it can be recreated with the same address range. Thus, it is possible that list can be different one.
```
            Ref<RunLoop::DispatchTimer> timer = RunLoop::current().dispatchAfter(timeout, [this, ptr, waiter = waiter.copyRef()] {
                if (RefPtr<WaiterList> list = findList(ptr)) {
                    Locker listLocker { list->lock };
                    if (waiter->ticket()) {
                        if (waiter->isOnList()) {
                            notifyWaiterImpl(listLocker, list->take(listLocker, waiter), ResolveResult::Timeout);
                            return;
                        }
                    }
                }
                ASSERT(!waiter->isOnList());
                if (waiter->ticket())
                    notifyWaiterImpl(NoLockingNecessary, waiter, ResolveResult::Timeout);
            });
```